numpyro/svi.py

```diff
@@ -52,32 +58,43 @@ def init_fn(rng, model_args=(), guide_args=(), params=None):
             guide_init = substitute(guide_init, params)
         guide_trace = trace(guide_init).get_trace(*guide_args, **kwargs)
         model_trace = trace(model_init).get_trace(*model_args, **kwargs)
+        inv_transforms = {}
         for site in list(guide_trace.values()) + list(model_trace.values()):
             if site['type'] == 'param':
-                params[site['name']] = site['value']
-        return optim_init(params)
+                constraint = site['kwargs'].pop('constraint', constraints.real)
+                transform = biject_to(constraint)
+                inv_transforms[site['name']] = transform
+                params[site['name']] = transform.inv(site['value'])
 
-    def update_fn(i, opt_state, rng, model_args=(), guide_args=()):
+        def constrain_fn(inv_transforms, params):
+            return {k: inv_transforms[k](v) for k, v in params.items()}
+
+        return optim_init(params), jax.partial(constrain_fn, inv_transforms)
+
+    def update_fn(i, rng, opt_state, constrain_fn, model_args=(), guide_args=()):
```

Comment:
The one disadvantage to this interface is that `constrain_fn` is the only static argument to `update_fn` and we will need to mark this as a static arg for the jit to work.

Reply:
I think we can simplify this by making `constrain_fn` a local variable of `svi`, and define it in `svi_init`.

Reply:
Yes, I think it is better to remove `constrain_fn` as args here. Will make the change.